elastic/kibana
# [GS] Allows apps to register searchable keywords for Global Search

Resolves https://github.com/elastic/kibana/issues/76778

## Summary

Allows applications to register keywords as search terms for Global Search. 
Since keywords are not data, they are have been added to the `App` type as a `meta` property. This PR moves `searchDeepLinks` into the new `meta` property for the same reason (they are not data).

```
core.application.register({
  id: 'myApp',
  title: 'Translated title,
  mount,
  meta: {
     keywords: ['translated keyword1', 'translated keyword2']
	 searchDeepLinks: [
     	{ id: 'sub1', title: 'Sub1', path: '/sub1', meta: { keywords: ['subpath1'] } },
   	 ],
  }
})
```

Scoring by keywords implements the same algorithm as searching for applications by title and weights the scores in favor of matching by title (keywords score * 0.8). The weighting factor is an arbitrary constant.

This PR makes no assumptions on what keywords to assign to apps, rather leaving it up to application developers and solutions to choose their own.

## How to test this

The unit tests for `get_app_info`, e.g. https://github.com/elastic/kibana/pull/85686/files#diff-5d4a779847b1ac5095589e30e25778e35a3350c09b4c6d9bdfcca0b22bc820fbR122 show how the keywords are added to an app and the unit tests for `get_app_results`, e.g. https://github.com/elastic/kibana/pull/85686/files#diff-3656691fb1cb333c33264ea83656f59878c49b13d7c3cdda8399ddbcc4e7d03eR113 show how scoring by keywords works. 

To test this 'live', you'll need to add keywords or add a new plugin that has keywords and, optionally, deep links (see https://github.com/elastic/kibana/pull/85686/files#diff-efac386db0738f497e14a9772c5da15dde935e474ee2dc3b62fe5fca2e0e92d8R247 as an example), after pulling the code.

### Example ###
Plugin: Uptime
Modified registration:

```ts
    core.application.register({
      id: PLUGIN.ID,
      euiIconType: 'logoObservability',
      order: 8400,
      title: PLUGIN.TITLE,
      category: DEFAULT_APP_CATEGORIES.observability,
      meta: {
        keywords: ['awake'], // --> add your keyword(s)
      },
      mount: async (params: AppMountParameters) => {
        const [coreStart, corePlugins] = await core.getStartServices();

        const { renderApp } = await import('./render_app');

        return renderApp(coreStart, plugins, corePlugins, params);
      },
    });
```
Run the code and search for 'awake'. This should show a direct link to 'Uptime'

<img width="696" alt="Uptime_awake_keyword" src="https://user-images.githubusercontent.com/11909450/102521642-6d4a1680-4052-11eb-93c8-9f9a1f604272.png">

Plugin: APM
Modified registration:

```ts
core.application.register({
      id: 'apm',
      title: 'APM',
      order: 8300,
      euiIconType: 'logoObservability',
      appRoute: '/app/apm',
      icon: 'plugins/apm/public/icon.svg',
      category: DEFAULT_APP_CATEGORIES.observability,
      meta: {
        keywords: ['application monitoring', 'performance'], // --> add your keyword(s)
      },

      async mount(params: AppMountParameters<unknown>) {
        ...
      },
    });
```

Searching for 'performance' or 'application monitoring' or any shorter version of these keywords will show a direct link to the 'APM' app:

<img width="697" alt="APM_performance_keyword" src="https://user-images.githubusercontent.com/11909450/102521830-a2566900-4052-11eb-9ba8-09a0ff16ef0e.png">

Keywords in deep links work similarly.

## Release Notes
Kibana's navigation search now implements keyword registration for applications, allowing you to search for applications by keywords as well as by title.

### Checklist

Delete any items that are not applicable to this PR.


### For maintainers
